File: 40-combination-sum-ii/combination-sum-ii.cpp
class Solution {
public:
    // Global
    vector<vector<int>> ans;
    vector<int> temp;

    void solve(vector<int> &candi, int target, int index, int sum){
        if (sum == target){
            ans.push_back(temp);
            return ;
        }

        if (sum > target || index == candi.size()) return ;

        temp.push_back(candi[index]);
        solve(candi, target, index + 1, sum+candi[index]);
        temp.pop_back();
        while(index+1 < candi.size() && candi[index] == candi[index+1]){
            index += 1;
        }
        solve(candi, target, index + 1, sum);
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        solve(candidates, target, 0, 0);
        return ans;
    }
};

File: 40-combination-sum-ii/combination-sum-ii.go
func solve(cand []int, target, sum, idx int, temp *[]int, ans *[][]int) {
    if sum == target {
        sol := make([]int, len(*temp))
        copy(sol, *temp)
        *ans = append(*ans, sol)
        return
    }
    
    if sum > target || idx >= len(cand) {
        return 
    }

    for i := idx; i < len(cand); i++ {
        if i > idx && cand[i] == cand[i-1] {
            continue
        }

        if sum + cand[i] > target {
            continue
        }

        *temp = append(*temp, cand[i])
        solve(cand, target, sum + cand[i], i+1, temp, ans)
        *temp = (*temp)[:len(*temp)-1]
    }
}

func combinationSum2(candidates []int, target int) [][]int {
    var temp []int
    var ans [][]int
    sort.Ints(candidates)
    solve(candidates, target, 0, 0, &temp, &ans)
    return ans
}